pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,578 @@
+			info = newServiceInfo(serviceName)
+			info.portal.ip = serviceIP
+			info.portal.port = servicePort.Port
+			info.portal.protocol = servicePort.Protocol
+			info.deprecatedPublicIPs = service.Spec.DeprecatedPublicIPs
+			// Deep-copy in case the service instance changes
+			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
+			info.sessionAffinityType = service.Spec.SessionAffinity
+			proxier.serviceMap[serviceName] = info
+
+			glog.V(4).Infof("info: %+v", info)
+		}
+	}
+
+	for name, info := range proxier.serviceMap {
+		// Check for servicePorts that were not in this update and have no endpoints.
+		// This helps prevent unnecessarily removing and adding services.
+		if !activeServices[name] && info.endpoints == nil {
+			glog.V(1).Infof("Removing service %q", name)
+			delete(proxier.serviceMap, name)
+		}
+	}
+
+	if err := proxier.syncProxyRules(); err != nil {
+		glog.Errorf("Failed to sync iptables rules: %v", err)
+	}
+}
+
+// OnEndpointsUpdate takes in a slice of updated endpoints.
+func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	glog.V(4).Infof("Received endpoints update notice: %+v", allEndpoints)
+	registeredEndpoints := make(map[proxy.ServicePortName]bool) // use a map as a set
+
+	// Update endpoints for services.
+	for i := range allEndpoints {
+		svcEndpoints := &allEndpoints[i]
+
+		// We need to build a map of portname -> all ip:ports for that
+		// portname.  Explode Endpoints.Subsets[*] into this structure.
+		portsToEndpoints := map[string][]hostPortPair{}
+		for i := range svcEndpoints.Subsets {
+			ss := &svcEndpoints.Subsets[i]
+			for i := range ss.Ports {
+				port := &ss.Ports[i]
+				for i := range ss.Addresses {
+					addr := &ss.Addresses[i]
+					portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortPair{addr.IP, port.Port})
+				}
+			}
+		}
+
+		for portname := range portsToEndpoints {
+			svcPort := proxy.ServicePortName{types.NamespacedName{svcEndpoints.Namespace, svcEndpoints.Name}, portname}
+			state, exists := proxier.serviceMap[svcPort]
+			if !exists || state == nil {
+				state = newServiceInfo(svcPort)
+				proxier.serviceMap[svcPort] = state
+			}
+			curEndpoints := []string{}
+			if state != nil {
+				curEndpoints = state.endpoints
+			}
+			newEndpoints := flattenValidEndpoints(portsToEndpoints[portname])
+
+			if len(curEndpoints) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpoints), newEndpoints) {
+				glog.V(1).Infof("Setting endpoints for %s to %+v", svcPort, newEndpoints)
+				state.endpoints = newEndpoints
+			}
+			registeredEndpoints[svcPort] = true
+		}
+	}
+	// Remove endpoints missing from the update.
+	for service := range proxier.serviceMap {
+		if _, exists := registeredEndpoints[service]; !exists {
+			glog.V(2).Infof("Removing endpoints for %s", service)
```

Comment:
Looking into it.

On Wed, Aug 12, 2015 at 1:09 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36827064>
> :
>
> > +			if state != nil {
> > +				curEndpoints = state.endpoints
> > +			}
> > +			newEndpoints := flattenValidEndpoints(portsToEndpoints[portname])
> > +
> > +			if len(curEndpoints) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpoints), newEndpoints) {
> > +				glog.V(1).Infof("Setting endpoints for %s to %+v", svcPort, newEndpoints)
> > +				state.endpoints = newEndpoints
> > +			}
> > +			registeredEndpoints[svcPort] = true
> > +		}
> > +	}
> > +	// Remove endpoints missing from the update.
> > +	for service := range proxier.serviceMap {
> > +		if _, exists := registeredEndpoints[service]; !exists {
> > +			glog.V(2).Infof("Removing endpoints for %s", service)
>
> I removed a service and saw this log line twice. Not sure why. Can you
> repro?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36827064>
> .
>